### PR TITLE
Add `direction_matrix` to `ImageData`

### LINKED ISSUE #5066: ImageData.points doesn't handle DirectionMatrix
### Describe the bug, what's wrong, and what you expected.

It seems that the ImageData.points property is not handling the image DirectionMatrix correctly. In fact, it has been ignored.

### Steps to reproduce the bug.

```python

import pyvista as pv
import numpy as np

img = pv.ImageData(dimensions=(2,3,4), spacing=(0.5,1.0,1.5), origin=(0., 0.5, 1.0))
img.SetDirectionMatrix((-1, 0, 0, 0, 1, 0, 0, 0, 1))

img.points
#array([[0. , 0.5, 1. ],
#       [0.5, 0.5, 1. ],
#       [0. , 1.5, 1. ],
#       [0.5, 1.5, 1. ],
#       [0. , 2.5, 1. ],
#       [0.5, 2.5, 1. ],
#       [0. , 0.5, 2.5],
#       [0.5, 0.5, 2.5],
#       [0. , 1.5, 2.5],
#       [0.5, 1.5, 2.5],
#       [0. , 2.5, 2.5],
#       [0.5, 2.5, 2.5],
#       [0. , 0.5, 4. ],
#       [0.5, 0.5, 4. ],
#       [0. , 1.5, 4. ],
#       [0.5, 1.5, 4. ],
#       [0. , 2.5, 4. ],
#       [0.5, 2.5, 4. ],
#       [0. , 0.5, 5.5],
#       [0.5, 0.5, 5.5],
#       [0. , 1.5, 5.5],
#       [0.5, 1.5, 5.5],
#       [0. , 2.5, 5.5],
#       [0.5, 2.5, 5.5]])

np.array([img.GetPoint(i) for i in range(img.number_of_points)])
#array([[ 0. ,  0.5,  1. ],
#       [-0.5,  0.5,  1. ],
#       [ 0. ,  1.5,  1. ],
#       [-0.5,  1.5,  1. ],
#       [ 0. ,  2.5,  1. ],
#       [-0.5,  2.5,  1. ],
#       [ 0. ,  0.5,  2.5],
#       [-0.5,  0.5,  2.5],
#       [ 0. ,  1.5,  2.5],
#       [-0.5,  1.5,  2.5],
#       [ 0. ,  2.5,  2.5],
#       [-0.5,  2.5,  2.5],
#       [ 0. ,  0.5,  4. ],
#       [-0.5,  0.5,  4. ],
#       [ 0. ,  1.5,  4. ],
#       [-0.5,  1.5,  4. ],
#       [ 0. ,  2.5,  4. ],
#       [-0.5,  2.5,  4. ],
#       [ 0. ,  0.5,  5.5],
#       [-0.5,  0.5,  5.5],
#       [ 0. ,  1.5,  5.5],
#       [-0.5,  1.5,  5.5],
#       [ 0. ,  2.5,  5.5],
#       [-0.5,  2.5,  5.5]])

```

### System Information

```shell
Date: Thu Oct 12 14:37:23 2023 CEST

                OS : Linux
            CPU(s) : 8
           Machine : x86_64
      Architecture : 64bit
               RAM : 15.0 GiB
       Environment : Python
       File system : ext4
        GPU Vendor : AMD
      GPU Renderer : RENOIR (renoir, LLVM 15.0.7, DRM 3.49, 6.2.0-34-generic)
       GPU Version : 4.6 (Core Profile) Mesa 23.0.4-0ubuntu1~22.04.1
  MathText Support : False

  Python 3.9.13 (main, May 30 2022, 12:53:10)  [GCC 11.2.0]

           pyvista : 0.42.3
               vtk : 9.2.6
             numpy : 1.23.4
        matplotlib : 3.5.2
            scooby : 0.5.12
             pooch : v1.6.0
            pillow : Module not found
           imageio : 2.19.3
           IPython : 8.4.0
        ipywidgets : 7.7.0
             scipy : 1.10.1
              tqdm : 4.64.0
            meshio : 5.3.4
        jupyterlab : 3.4.2
             trame : Version unknown
      trame_client : Version unknown
      trame_server : 2.9.1
         trame_vtk : Version unknown
     trame_vuetify : Version unknown
      nest_asyncio : Version unknown
```

### Screenshots

_No response_

### PR DESCRIPTION
**Pull request by user27182**
### Overview

- Add `direction_matrix` property
- Use it in `points` to fix #5066
- ~Update `bounds` to support direction as well~ `GetBounds` already considers direction matrix internally 
- ~Add `index_to_physical_matrix` and `physical_to_index_matrix` to move between index and physical space~ split into #6614 instead

### CODE REVIEW

**Review by banesullivan — APPROVED** [review]
Looks great!

File: pyvista/core/grid.py [hunk 1]
```diff
@@ -670,7 +673,12 @@ def points(self) -> NumpyArray[float]:
         y = np.insert(np.cumsum(np.full(ny, dy)), 0, 0.0) + oy
         z = np.insert(np.cumsum(np.full(nz, dz)), 0, 0.0) + oz
         xx, yy, zz = np.meshgrid(x, y, z, indexing='ij')
-        return np.c_[xx.ravel(order='F'), yy.ravel(order='F'), zz.ravel(order='F')]
+        points = np.c_[xx.ravel(order='F'), yy.ravel(order='F'), zz.ravel(order='F')]
+
+        direction = self.direction_matrix
+        if not np.array_equal(direction, np.eye(3)):
+            return pyvista.Transform().rotate(direction, point=self.origin).apply(points)
```
Comment (user27182) [inline pyvista/core/grid.py:680]: Small optimization, no need to make a copy here

```suggestion
            return pyvista.Transform().rotate(direction, point=self.origin).apply(points, copy=False)
```

**Review by tkoyama010 — APPROVED** [review]

**Comment by tkoyama010** [comment]
pre-commit.ci autofix

**Comment by user27182** [comment]
pre-commit.ci autofix

**Review by tkoyama010 — APPROVED** [review]

**Pull request merged**